### ISSUE TITLE
How can I get each account key type?

### ISSUE THREAD

**Issue by MegaSolar** [klaytn/klaytn#433]
Hi, I just trying to send transaction through EN. 
I have some questions about account key.

I have made account through EN ipc. 
and type ```klay.getAccountKey("0xaccountaddr","latest") ``` returns null. 

How can I make account key each different purpose?

I have read https://docs.klaytn.com/klaytn/design/accounts#account-key. 
and I have searched https://scope.klaytn.com/ for search some account key type. 
Some accounts return role based key and some accounts return null. 
I don't know why &  makes me confused. 

in short. 

1. How can I make each account key(accountkeyLegacy, accountkeyRolebase ..etc) through RPC in local EN? 
2. Why my account returns null when I use ```klay.getAccountKey```?

Thanks in advance.

**Comment by kjhman21** [comment]
Thanks for filing an issue. @aidan-kwon, could you please answer to this question?

**Comment by MegaSolar** [comment]
@aidan-kwon ping

**Comment by aidan-kwon** [comment]
Answer 1-1. At first, you should create a new account key to be updated. To create an accoun key, you can use node API (https://docs.klaytn.com/bapp/json-rpc/api-references/klay/account#klay_encodeaccountkey) or caver SDKs (https://docs.klaytn.com/bapp/sdk/caver-js/getting-started#accountkey or https://docs.klaytn.com/bapp/sdk/caver-java/getting-started#using-various-account-key-type) 
Answer 1-2. You should update your account with the new account key through AccountUpdate transaction. For more information about AccountUpdate tx, https://docs.klaytn.com/klaytn/design/transactions/basic#txtypeaccountupdate 

Answer 2. It should return some values if the account is created. @MegaSolar Can I know the exact behavior you did in "I have made account through EN ipc."? 
Generating a key pair doesn't mean generating an account in the Klaytn network. The address should receive KLAY or generate a transaction to be created in the global state (which means the storage of Klaytn nodes).

**Comment by MegaSolar** [comment]
@aidan-kwon Thanks for reply. 
As I understand, I need x,y factors for creating ```AccountKeyPublic``` & ```RoleBasedAccountKey```. 
How can I get x,y factor for these through  json rpc?

**Comment by MegaSolar** [comment]
According to Answer 2.
Klaytn version  : Klaytn/v1.3.0/linux-amd64/go1.13.3
Network : baobab
Method in console ipc 
```
personal.newAccount(pass)
# then rightaway type below method without any transaction on new account. 
klay.getAccountKey(newaccount,"latest")
null
```
When I try aboves, returned ``null``.

**Comment by aidan-kwon** [comment]
> As I understand, I need x,y factors for creating `AccountKeyPublic` & `RoleBasedAccountKey` when using `encodeaccountkey` method.
> How can I get x,y factor for these through json rpc?

The first half of a public key is x and the second half of a public key is y when the public key is 64 bytes hex. Currently, Klyatn node APIs don't show the x, y values (or public key) kindly from a newly generated private key. 

If you don't have any other tools to get a public key from your keystore file, I recommend using Klaytn SDK called `caver`. You can see kind instructions to update your account key in https://docs.klaytn.com/bapp/sdk/caver-js/getting-started#accountforupdate. Even you can generate a new key pair and print the public key of the key pair. Please,

**Comment by aidan-kwon** [comment]
> just only I can have account key type is "1" & key object is nothing?
> (e.g

`AccountKeyLegacy` type account key doesn't have key values but has key type 1.
All user account (EOA) didn't update their account key yet has `AccountKeyLegacy` type account key as default.
Refer https://docs.klaytn.com/klaytn/design/accounts#accountkeylegacy.

**Comment by aidan-kwon** [comment]
> >personal.newAccount(pass)

The operation generates a key pair in your local machine but didn't share the account to other nodes since the account doesn't generate any transactions. 
As I mentioned before, the address should receive KLAY or generate a transaction to be created in the global state (which means the storage of Klaytn nodes).

**Comment by MegaSolar** [comment]
@aidan-kwon Thanks for reply. 
I tried to account key update as you said, then this was successful. 

I have another question to make tx type ```TxTypeFeeDelegatedValueTransfer```. 

I encoded among role based account key only fee delegation role key (key[2] I mean). 
then I signed transaction, I got result. ( I use klay_signTransaction)
then I try to make tx to use ``klay_sendTransactionAsFeePayer``. I got error below. 
```
{"jsonrpc":"2.0","id":69,"error":{"code":-32000,"message":"unknown account"}}

```

then I have used ```sendRawTransaction``` method for solve this. 
result is different, but I got error either. 

```
{"jsonrpc":"2.0","id":1,"error":{"code":-32000,"message":"invalid fee payer"}}

````
even those 3 accounts ;from , to ,feepayer account that existed in baobab network. 
If you don't mind could you check this acc; **0x633196dfbd097a38d94ec571bfdeb41eff017909**? 
I don't know why this error occurred .

**Comment by aidan-kwon** [comment]
@MegaSolar I think you misunderstood something about the fee-delegation.

At first, fee-delegation transactions including `TxTypeFeeDelegatedValueTransfer` require two signatures from a sender and a fee payer. After you make the transaction, you should sign the tx with the sender's private key through `klay_signTransaction` and signs the sender signed tx with the fee payer's private key through `klay_signTransactionAsFeePayer`. Then, you can send the doubly signed tx to the Klaytn network through `klay_sendRawTransaction`.

Second, you don't need to have a role-based key with RoleFeePayer to send `TxTypeFeeDelegatedValueTransfer` transaction. The account who has any account key except FailKey can sign the tx as a fee payer. 

Let's suppose that you have 3 accounts with LegacyKey (default key): A(sender), B(receiver), C(fee payer). After generating a fee-delegating tx which has proper field values (from: A, to: B, feePayer: C), you just need to sign the tx with A's key through `klay_signTransaction` and signs again with C's key through `klay_signTransactionAsFeePayer`. Of course, the node receiving the API call should have both private keys.

**Comment by MegaSolar** [comment]
@aidan-kwon Thanks! Its helpful!
Then how can I get v, r, s for using ```klay_signTransactionAsFeePayer```?

**Comment by aidan-kwon** [comment]
The signature consists of v, r, s.
You can get v, r, s of sender's signature using `klay_signTransaction`.
You can get v, r, s of fee payer's signature using `klay_signTransactionAsFeePayer`.

**Comment by MegaSolar** [comment]
@aidan-kwon Thanks now I tried one by one as your instruction. 
please look at below curl command 

```
curl -H "Content-Type: application/json" --data '{"jsonrpc": "2.0", "method": "klay_signTransaction", "params": [{"from": "0xef21b226b42a30df68d3ccd9ce3213792dc882f2", "to": "0x9ad19506a71a87675d88d1628c48363c0cf534b6", "value": "0x16345785D8A0000"}], "id": 59}' http://127.0.0.1:8551 | python -m json.tool
```

result 

```
{
    "id": 59,
    "jsonrpc": "2.0",
    "result": {
        "raw": "0xf86f0d8505d21dba0083015f90949ad19506a71a87675d88d1628c48363c0cf534b688016345785d8a0000808207f6a0d18a74b3b3cd102947b14c4f887cf4db127d33d3ae4c24070157e7233640a51ba0550d9c7f65c262b57f9f1374bb600578060d575549f7f6d35beed6e5aa95e135",
        "tx": {
            "gas": "0x15f90",
            "gasPrice": "0x5d21dba00",
            "hash": "0xbc3d8c062c04e71f6c17f4aff5647e6a387273a21d9acc90139c58304dbdf7bc",
            "input": "0x",
            "nonce": "0xd",
            "signatures": [
                {
                    "R": "0xd18a74b3b3cd102947b14c4f887cf4db127d33d3ae4c24070157e7233640a51b",
                    "S": "0x550d9c7f65c262b57f9f1374bb600578060d575549f7f6d35beed6e5aa95e135",
                    "V": "0x7f6"
                }
            ],
            "to": "0x9ad19506a71a87675d88d1628c48363c0cf534b6",
            "value": "0x16345785d8a0000"
        }
    }
}
```

then I sign to use klay_signTransactionAsFeePayer 
```
curl -H "Content-Type: application/json" --data '{"jsonrpc": "2.0", "method": "klay_signTransactionAsFeePayer", "params": [{"from": "0xef21b226b42a30df68d3ccd9ce3213792dc882f2", "to": "0x9ad19506a71a87675d88d1628c48363c0cf534b6", "value": "0x16345785D8A0000","feePayer":"0x633196dfbd097a38d94ec571bfdeb41eff017909","Signatures":[{"V":"0x7f6","R":"0xd18a74b3b3cd102947b14c4f887cf4db127d33d3ae4c24070157e7233640a51b","S":"0x550d9c7f65c262b57f9f1374bb600578060d575549f7f6d35beed6e5aa95e135"}],"typeInt":9}], "id": 59}' http://127.0.0.1:8551 | python -m json.tool
```

result 

```
{
    "id": 59,
    "jsonrpc": "2.0",
    "result": {
        "raw": "0x09f8e50d8505d21dba0083015f90949ad19506a71a87675d88d1628c48363c0cf534b688016345785d8a000094ef21b226b42a30df68d3ccd9ce3213792dc882f2f847f8458207f6a0d18a74b3b3cd102947b14c4f887cf4db127d33d3ae4c24070157e7233640a51ba0550d9c7f65c262b57f9f1374bb600578060d575549f7f6d35beed6e5aa95e13594633196dfbd097a38d94ec571bfdeb41eff017909f847f8458207f6a0f9438ad145c5316baa55826156ac3b0283a8a847edde337be0c3027fbc83d3c9a025bd0b1f5f4eb7265d589ba10aeddbfee28e6c27c83ea83c23b263be0fb22113",
        "tx": {
            "feePayer": "0x633196dfbd097a38d94ec571bfdeb41eff017909",
            "feePayerSignatures": [
                {
                    "R": "0xf9438ad145c5316baa55826156ac3b0283a8a847edde337be0c3027fbc83d3c9",
                    "S": "0x25bd0b1f5f4eb7265d589ba10aeddbfee28e6c27c83ea83c23b263be0fb22113",
                    "V": "0x7f6"
                }
            ],
            "from": "0xef21b226b42a30df68d3ccd9ce3213792dc882f2",
            "gas": "0x15f90",
            "gasPrice": "0x5d21dba00",
            "hash": "0xb74c8949e4a61deff62a90cf24dc05b9b4702f4048295a1fe7796e5ce4750ab8",
            "nonce": "0xd",
            "signatures": [
                {
                    "R": "0xd18a74b3b3cd102947b14c4f887cf4db127d33d3ae4c24070157e7233640a51b",
                    "S": "0x550d9c7f65c262b57f9f1374bb600578060d575549f7f6d35beed6e5aa95e135",
                    "V": "0x7f6"
                }
            ],
            "to": "0x9ad19506a71a87675d88d1628c48363c0cf534b6",
            "type": "TxTypeFeeDelegatedValueTransfer",
            "typeInt": 9,
            "value": "0x16345785d8a0000"
        }
    }
}
```
And sendRawTransaction
```
curl -H "Content-Type: application/json" --data '{"jsonrpc": "2.0", "method": "klay_sendRawTransaction", "params": ["0x09f8e50d8505d21dba0083015f90949ad19506a71a87675d88d1628c48363c0cf534b688016345785d8a000094ef21b226b42a30df68d3ccd9ce3213792dc882f2f847f8458207f6a0d18a74b3b3cd102947b14c4f887cf4db127d33d3ae4c24070157e7233640a51ba0550d9c7f65c262b57f9f1374bb600578060d575549f7f6d35beed6e5aa95e13594633196dfbd097a38d94ec571bfdeb41eff017909f847f8458207f6a0f9438ad145c5316baa55826156ac3b0283a8a847edde337be0c3027fbc83d3c9a025bd0b1f5f4eb7265d589ba10aeddbfee28e6c27c83ea83c23b263be0fb22113"], "id": 59}' http://127.0.0.1:8551 | python -m json.tool
```

result has error about signature. 

```
{
    "error": {
        "code": -32000,
        "message": "invalid transaction v, r, s values of the sender"
    },
    "id": 59,
    "jsonrpc": "2.0"
}
```

what is the reason?

**Comment by MegaSolar** [comment]
Q 2. 
then How can I use ```klay_sendTransactionAsFeePayer``` ? 
Before use ```klay_sendTransactionAsFeePayer```, I have to ```klay_signTransaction`` and get v,r,s for ```klay_sendTransactionAsFeePayer```?

**Comment by aidan-kwon** [comment]
Answer 1. You used `klay_signTransactionAsFeePayer` and `klay_sendRawTransaction` correctly, but put incorrect parameters for `klay_signTransaction`.  You should put the type int value also to call `klay_signTransaction`. 

Try the command below and call `klay_signTransactionAsFeePayer` and `klay_sendRawTransaction` as you did. 
```
curl -H "Content-Type: application/json" --data '{"jsonrpc": "2.0", "method": "klay_signTransaction", "params": [{"from": "0xef21b226b42a30df68d3ccd9ce3213792dc882f2", "to": "0x9ad19506a71a87675d88d1628c48363c0cf534b6", "value": "0x16345785D8A0000","feePayer":"0x633196dfbd097a38d94ec571bfdeb41eff017909"}],"typeInt":9}], "id": 59}' http://127.0.0.1:8551 | python -m json.tool
```

**Comment by aidan-kwon** [comment]
Answer 2. 
`klay_sendTransactionAsFeePayer` == `klay_signTransactionAsFeePayer` + `klay_sendRawTransaction`

1. Get the sender signature (v, r, s) with `klayt_signTransaction` or other methods.
2. Call `klay_sendTransactionAsFeePayer` with correct parameters. (The same parameter you did to call `klay_signTransactionAsFeePayer`.
3. That it! You don't need to call `klay_sendRawTransaction`.

**Comment by MegaSolar** [comment]
@aidan-kwon Wow. 
Such a nice instruction. Thanks!

**Comment by MegaSolar** [comment]
@aidan-kwon 
I think your curl command a bit strange. 

```
curl -H "Content-Type: application/json" --data '{"jsonrpc": "2.0", "method": "klay_signTransaction", "params": [{"from": "0xef21b226b42a30df68d3ccd9ce3213792dc882f2", "to": "0x9ad19506a71a87675d88d1628c48363c0cf534b6", "value": "0x16345785D8A0000","feePayer":"0x633196dfbd097a38d94ec571bfdeb41eff017909"}],"typeInt":9}], "id": 59}' http://127.0.0.1:8551 | python -m json.tool
```

Your curly bracket and square bracket position a bit weird. 

I think below command is what you wanted to said, is it? 

```
 curl -H "Content-Type: application/json" --data '{"jsonrpc": "2.0", "method": "klay_signTransaction", "params": [{"from": "0xef21b226b42a30df68d3ccd9ce3213792dc882f2", "to": "0x9ad19506a71a87675d88d1628c48363c0cf534b6", "value": "0x16345785D8A0000","feePayer":"0x633196dfbd097a38d94ec571bfdeb41eff017909","typeInt":9}], "id": 59}' http://127.0.0.1:8551 | python -m json.tool
```

**Comment by aidan-kwon** [comment]
@MegaSolar Yes, you are right. There was a typo in the command.

**Comment by MegaSolar** [comment]
@aidan-kwon Wow even in weekend you are working. 
please take a look updated comment above?

**Comment by MegaSolar** [comment]
@aidan-kwon that's why.. I got it . 
so I can't use role-based key in the cypress EN through jsonrpc? 
I have to use caver-js / caver-java ?

**Comment by MegaSolar** [comment]
@aidan-kwon OK. I got it. 
Thanks for everything!

**Comment by MegaSolar** [comment]
@aidan-kwon And So much appreciated even in the weekend. 
God bless you & Klaytn team!

**Comment by aidan-kwon** [comment]
> so I can't use role-based key in the cypress EN through jsonrpc?
> I have to use caver-js / caver-java ?

Yes. Klaytn node wallet will be updated later.
@MegaSolar Thanks. Have a good weekend!
